### ISSUE TITLE
TG2 - Note Test dependencies and/or workflows

### ISSUE THREAD

**Issue by Tasilee** [tdwg/bdq#186]
Beyond the agreed VALIDATE - AMEND - RE-VALIDATE, we will have a subset of tests that are dependent on a prior sequence of other tests having been run. We have a dependency worksheet that I have now updated: https://docs.google.com/spreadsheets/d/1cWxf1vABLHuO9g1NhpjYCHU_bXWkiV1NCenpx59pkkI but it will need editing given some changes to tests such as #147?

**Comment by tucotuco** [comment]
It's good to have the spreadsheet with all of the tests in one place to peruse, but I am concerned that the spreadsheet does not allow us to define the sequence - only the dependencies. In addition, dependency could be a little confusing, partly due to the question of which is dependent on which. Does it mean, I shouldn't be run until those in my dependency list are run? Or does it mean, those in my dependency list depend on me? I guess from the entries that it should be the former, but for the uninitiated, that might not be enough. In any case, the order of tests in a given sequence is what is really important, especially since it could be extremely useful to run some of them before and after a sequence of tests. The way it is now, it seems like there is some circularity in dependencies. The chain that includes considerations of coordinates and country codes is a good example, and a good one to make a model from.

**Comment by Tasilee** [comment]
Thanks @tucotuco. I just wanted the work that Arthur in particular put into that spreadsheet not to be lost in the path toward sequences. Given what we know about dependencies, it was always going to require something equivalent to a decision tree as @pzermoglio had done for date/time.

I agree that the spatial 'tests' (we need a better name BTW) would be a good place to start.

**Comment by tucotuco** [comment]
Is anyone in particular tasked with trying to make trees? Should we use
Paula's model (where is it)? Should we distribute the task? Or better that
someone does all the first drafts and then we all review?

**Comment by Tasilee** [comment]
This would be a nice AI application: Recursion and all that.

Regards @pzermoglio ’s diagram, I sent the XML file that Paula did from draw.io and was able to drag and drop it into that fine. I got that file from Paula’s original email to us. I’m on the iPad at the moment so will send link to the drawing when I get to a PC.

A decision tree would seem the easiest way to understand/communicate the sequences but another form may be more suited to a standard. Sharing load: Probably best as the overall task could be a lot for anyone. I’d prefer to focus on the test datasets in what time I have available, and as you all know, I am not one for detail :)

**Comment by Tasilee** [comment]
Can we agree that the tests are to be treated as independent, other than the overall process of VALIDATE-AMEND-VALIDATE? In other words, none of the tests are dependent on a workflow?

**Comment by Tasilee** [comment]
Title
Data Quality Tests and Assertions
Date version issued

Date created

Part of TDWG Standard

This version

Latest version

Previous version

Abstract (Lee)
There was a recognition at TDWG 2010 that a standard suite of ‘data quality’ tests applied to Darwin Core records would have extensive benefits to biodiversity-related data collectors and to subsequent users of those records. ‘Data quality’ is used in this document, as a generic term that is interpreted here as ‘fitness for use’, as the ‘quality’ of a data record depends on the context in which it is used.

The work commenced in 2014 by reviewing over 150 known ‘data quality’ tests used by GBIF, the ALA, iDigBio, CRIA, BISON  and others. The tests were classified either as ‘core’ if they fulfilled all basic criteria developed by the group, or ‘supplemental’ if they fulfilled most and could be useful in some contexts. Criteria for inclusion included coverage of key Darwin Core terms, ease of implementation and utility of the test response. Additional tests were added to the core group to fill gaps that were detected and the resulting 99 tests have been considered stable. 

The work on this standard is based on the Fitness for Use Framework of the Data Quality Task Group 1: Framework on Data Quality. The most significant component of the work on these tests were the specifications for the tests. These specifications provided the foundation for the concise description of the tests such that implementations would be straight-forward. The terms used in the specifications form a large component of the associated vocabulary generated by this group..

A suite of over 1100 Darwin Core records was developed and then applied to the implemented tests. The results helped to refine the tests.

Not dependent on workflow
Contributors
Lee Belbin, Arthur Chapman, Paul Morris, John Wieczorek, Paula Zermoglio and Alex Thompson.


Creator
TDWG Biodiversity Data Quality Interest Group: Task Group 2 (Data Quality Tests and Assertions).

Bibliographic citation

1 Introduction (Informative) (Later) (John)

This document contains  

1.1. Status of the content of this document

Section 1 is informative.  Section 2 is informative.  Portions of sections 3-6 are informative, and portions are normative.  These are marked as such.

1.2 RFC 2119 keywords

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

1.3 Namespace abbreviations (Paul and John)

The following namespace abbreviations are used in this document:

Abbreviation	IRI
Abbreviation	IRI
bdq:	https://rs.tdwg.org/bdq/terms/
bdqffdq:	https://rs.tdwg.org/bdq/ffdq/
dwc:	https://rs.tdwg.org/dwc/terms/
dwciri:	https://rs.tdwg.org/dwc/iri/
dc:	https://purl.org/dc/elements/1.1/
dcterms:	https://purl.org/dc/terms/

1.4 About the tests, their use and specifications. (Lee)

This standard is a specification of tests for the quality of biodiversity data, not a specification of the quality to which biodiversity data are expected to conform. ‘Fitness for use’ of a biodiversity data record will greatly depend on the use to which it is applied.

This standard consists of a framework for describing data quality, a set of vocabularies related to data quality in the biodiversity domain, the description of a set of tests of biodiversity data quality and a suite of test data that can be used to validate test implementations.   A user guide to interpretation of the results of the tests, and a guide to implementation of the tests are included.

This standard describes a suite of core tests, where ‘core’ implies that the tests are  informative, simple to implement, are mandatory for enhancements/amendments, have ‘power’ in that they will not likely result in 0% or 100% of all records fail or pass, are widely applicable across sub-disciplines, and/or may elevate an issue (e.g., no value for dcterms:license)

The tests are intended for use in quality control and quality assurance of biodiversity data.  They serve to assess the fitness for use of biodiversity occurrence data for uses where key components of Darwin Core encoded records are important for an application.   The scope of this core was established by an analysis of research needs for biodiversity data (a product of TDWG Biodiversity Data Quality Task Group 3 - Data Quality Use Cases).   Subsets of the tests MAY be used to assess the quality of other forms of biodiversity and other data.   Implementers MAY add additional tests based on the test specifications in this standard, to cover additional data quality needs associated with specific applications. 

The test specifications all assume that data are presented to the tests in structured form, with data elements identifiable as Darwin Core terms.  The test specifications are agnostic about whether the data are flat, or have a more complex structure.  The tests are agnostic about the form in which the data are stored or transported. The tests are also agnostic about uses for quality assurance, where output data are limited to those for which all Validations are Compliant, or for quality control, where the results of Validations, Measures, and Amendments can be used to improve the quality of the data for core purposes. The test specifications are agnostic about where in the life of biodiversity data they are being run.   Implementers may incorporate tests anywhere where they may be of utility, including into tools for data capture in the field, tools for data transcription from paper records, in data processing pipelines for ingest into a database of record, integrated into collections management systems and other systems for managing databases of record, in data aggregation pipelines and tools, by data aggregators, in tools for managing extended specimen network data, and in data processing pipelines for research on aggregated biodiversity data.

The core tests are all defined as single record-level tests, working on a set of data structured into Darwin Core terms. None of the tests rely on more than a single Darwin Core encoded record, but may require reference to external data such as standard vocabularies of terms or names. The test results may be accumulated across multiple records, for example 75% of the records do not have a valid value for dwc:basisOfRecord.

The tests are described by the following xx specifications:

Title: A standardised, human readable name of the test-assertion based on the template 
OUTPUTTYPE_TERMS_RESPONSE. These names were considered helpful for
human-human communication and to assist with code implementation, maintenance and searches. Example: VALIDATION_BASISOFRECORD_STANDARD.

GUID: A globally unique identifier which allows software to uniquely identify each test (and in combination with parameter values, allows for specification of the expectations for the behavior of a test implementation). Example: 42408a00-bf71-4892-a399-4325e2bc1fb8. 
 
Description: An english language brief description. Example: Does the value of dwc:basisOfRecord occur in bdq:sourceAuthority? 

Output Type: Tests have been classified into three Fitness for Use Framework classes;  VALIDATION (validates values in one or more Darwin Core terms),  AMENDMENT (an
improvement that will result in a change or addition to at least one Darwin Core term);
and MEASURE (returns a numeric value, for the tests described here; all values are in the form of the number of tests that conform to a criterion). Three tests are typed as ISSUE (flags a potential issue where data remains valid, a concept outside the FFU Framework). Example: VALIDATION. Three tests are typed as ISSUE (flags a potential problem). Example: POTENTIAL_ISSUE.

Darwin Core Class: The Darwin Core class that contains the Information Elements. Example: dwc:Taxon

Information Elements: The Darwin Core terms that the test takes as input. Example: dwc:basisOfRecord

Expected Response: A concise description of the logic of the test to clarify implementation. The Expected Response takes the form of (for a VALIDATION) of: EXTERNAL_PREREQUISITES_NOT_MET if <condition>; INTERNAL_PREREQUESITES_NOT_MET if <condition>; COMPLIANT if <condition>; otherwise NOT_COMPLIANT. Example: EXTERNAL_PREREQUISITES_NOT_MET if the bdq:sourceAuthority is not available; INTERNAL_PREREQUISITES_NOT_MET if dwc:basisOfRecord is EMPTY; COMPLIANT if the value of dwc:basisOfRecord is valid using the bdq:sourceAuthority; otherwise NOT_COMPLIANT

Data Quality Dimension: A test will focus on one of the following scenarios based on the Data Quality Framework: "Completeness" (the extent to which data elements are present and sufficient); "Conformance" (Conforms to a format, syntax, type, range, standard or to the own nature of the information element); "Consistency" (agreement among related information elements in the data); "Likeliness" (low probability that values are real); "Resolution" (is sufficient detail present in the value/s - a measure the granularity of the data).

Warning Type: The nature of the flag associated with the result of the test. Possible values are "Ambiguous", "Amended", "Incomplete", "Inconsistent", "Invalid", "Notification", "Report" and "Unlikely

Parameter(s): Parameters that modify the behavior of the test, along with default values or links to source authorities. Example: bdq:sourceAuthority	

Source Authority: A reference to the authority required by the test. Example: bdq:sourceAuthority default = "Darwin Core Terms" [https://dwc.tdwg.org/terms/#dwc:basisOfRecord]

Examples: A ’pass’ and a ‘fail’ example of test data. Example: [dwc:basisOfRecord="Taxon": Response.status=RUN_HAS_RESULT, Response.result=COMPLIANT, Response.comment="dwc:basisOfRecord matches a standard label of one of the Darwin Core classes"]
[dwc:basisOfRecord="Specimen": Response.status=RUN_HAS_RESULT, Response.result=NOT_COMPLIANT, Response.comment="dwc:basisOfRecord does not exactly match a standard label of one of the Darwin Core classes"]

Source: The origin of the concept of the test. Example: The Atlas of Living Australia

References: One or more publications that relate directly to the test. Example: GBIF Secretariat (2021). GBIF Backbone Taxonomy https://www.gbif.org/dataset/d7dddbf4-2cf0-4f39-9b2a-bb099caae36c)

Example Implementations (Mechanisms): A link to one or more agencies that have an
implementation of the test. Example: https://github.com/FilteredPush/event_date_qc

Link to Specification Source Code: A link to reference code set that
demonstrates the test. Example: https://github.com/FilteredPush/ event_date_qc/blob/
5f2e7b30f8a8076977b2a609e0318068db80599a/src/main/java/org/filteredpush/qc/date/
DwCEventDQ.java#L169

Notes: Additional comments that the Task Group believed necessary for an accurate
understanding of the test or issues that implementers needed to be aware of.
Example: For VALIDATION_COUNTRYCODE_CONSISTENT, “The country code determination service should be able to match the name of a country in the original language. This test will fail if there is leading or trailing whitespace or there are leading or trailing non-printing characters”.


Tests not dependent on workflow!?



1.5 Framework for describing data quality (Paul)
Included in this standard is a specification for
2 Guidance for Consumers of Data Quality Reports (Informative)

2.1 Introduction (Arthur)
An internationally agreed standard suite of core tests and resulting assertions can be used by all data providers, data collectors and data users to improve the quality of the data. This will allow for more appropriate and more accurate use of biodiversity data.
Other than data availability, ‘Data Quality’ is probably the most significant issue for users of biodiversity data and this is especially so for the research community. 
The tests will not correct all issues that exist with the data, but reports from the tests will identify issues that need to be addressed by users of the data. This may require the user to make decisions on the data - i.e., data that may need to be excluded, data that may need examining for possible improvement, and data that can be used as is. It is always the purview of the user to decide what data is of suitable quality for their use.

2.2 Biodiversity Data Quality Tests (Lee)
What are the important attributes of data quality for biodiversity data?  How are we proposing to assess data quality in the domain?  What is the scope of the tests?   What is out of scope for the tests.  How did we develop the test specifications? 

We identified four fundamental aspects of biodiversity-related data that we needed to cover by the tests: Name (taxonomic information); space (geographic location); time (temporal terms) and other (all other terms such as dwc:basisOfRecord). A record without a taxonomic name, a location or a date has limited value. Three tests in this standard specifically target records with no name, space and time values.

[John’s statement on calendars as a placeholder (https://github.com/tdwg/bdq/issues/76#issuecomment-1591985055): “I don't think we are in any position to posit a date after which the Gregorian calendar assumption is safe. It is still not safe today, it's just that its use for civil purposes has ever fewer exceptions as time goes on (so far). Making a statement about a particular date (other than the date of its origin) for a date of special mention necessarily has discriminatory implications. We do NOT want that.

In this particular issue, and perhaps in all others where this has come up, I do not see that the uncertainty associated with the date actually has anything to do with what we are testing. This test can't assess if a date is actually within a Gregorian date interval, except in special cases where the Julian and Gregorian calendars coincide, and even that is ignoring all other possible calendars. Instead, it is able to test that a date following the ISO 8601-1 date specification is within a range specified in that context. We can't effectively do anything else because Darwin Core doesn't even provide for stating the original calendar used - it's forcing people to use the Gregorian calendar without describing the responsibility for doing so and the consequences of not doing so. I think the place for awareness of the implications of dates with unknown calendars is in the Darwin Core date terms.]




https://github.com/tdwg/bdq/blob/master/tg2/core/TG2_tests.csv
https://github.com/tdwg/bdq/blob/master/tg2/core/TG2_tests.xml

[Include generated text]


3. Guidance for Implementers (Paul)
Should this go under 6.2

PLACEHOLDER: The value supplied for the parameter for the test is not an attribute of the data, it is an attribute of the Mechanism (of the system assessing the data quality). If we had included assertions about the validity values of parameters, they should only return external prerequisites not met, as they are assertions about externalities to the data and will change if the same data are run on the same test with a different configuration.

Tests not dependent on workflow!?

4 A Framework for Data Quality (Paul)

4.1 Introduction

4.2 Framework (Normative)

[Include generated text]

5. Vocabularies (Arthur)

5.1 Introduction
The included vocabulary (see Supplement 2) is of terms used for the Data Quality Tests and Assertions. Some definitions have been updated since the publication by Veiga et al. and where the change is significant, this is noted in the comments. Vocabulary terms include a column that indicates the Context in which the terms are used.
 
5.2 Vocabularies
SUPP 2

[Include generated text]

6 Test Specifications (Paul)

6.1 Introduction (Informative)

6.2 Implementation (Normative)

In order to be considered as compliant with this standard, an implementation MUST

Results from each test MUST be produced in the form Response.status, Response.result, and Response.comment, with one test producing one Response.   Results MAY include Response.qualifier (See 5.3).  The values of Response.status and Response.result MUST be those specified.   This standard is agnostic concerning data structures and serializations of a Response.   The standard is agnostic concerning internationalization and languages of labels applied to human readable presentations of values within a Response.

Implementations MAY run Validations, Measures, and Issues in a pre-Amendment phase, then run Amendments, then re-run the Validations, Measures, and Issues on the data with the proposed changes from the Amendments applied in a post-Amendment phase.   The use of pre-amendment, amendment, and post-amendment phases allows for measurement of how much acceptance of the changes proposed by the amendments would improve the quality of the data for core purposes.

For each core SingleRecord Validation, an implementation intended for Quality Assurance SHOULD include a corresponding MultiRecord Measure that counts the number of Response.result values that are COMPLIANT.   

6.3 Extension points (Normative)

A response MAY include a Response.qualifier.  This is intended as a place to include structured assertions concerning uncertainty in a response.


6.4 Core Test Specifications (Normative)

Label	Description	url
MEASURE_AMENDMENTS_PROPOSED	The number of distinct AMENDMENT tests that have a Response.status="AMENDED" for a given record.	https://api.github.com/repos/tdwg/bdq/issues/65
ISSUE_ANNOTATION_NOTEMPTY	Are there any annotations associated with the record?	https://api.github.com/repos/tdwg/bdq/issues/29
VALIDATION_BASISOFRECORD_NOTEMPTY	Is there a value in dwc:basisOfRecord?	https://api.github.com/repos/tdwg/bdq/issues/58
VALIDATION_BASISOFRECORD_STANDARD	Does the value of dwc:basisOfRecord occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/104
AMENDMENT_BASISOFRECORD_STANDARDIZED	Propose amendment to the value of dwc:basisOfRecord using bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/63
VALIDATION_CLASS_FOUND	Does the value of dwc:class occur at rank of Class in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/77
VALIDATION_CLASSIFICATION_CONSISTENT	Is the combination of higher classification taxonomic terms consistent using bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/123
AMENDMENT_COORDINATES_CONVERTED	Propose amendment to the value of dwc:geodeticDatum and potentially to dwc:decimalLatitude and/or dwc:decimalLongitude based on a conversion between spatial reference systems.	https://api.github.com/repos/tdwg/bdq/issues/43
VALIDATION_COORDINATES_COUNTRYCODE_CONSISTENT	Do the geographic coordinates fall on or within the boundaries of the territory given in dwc:countryCode or its Exclusive Economic Zone?	https://api.github.com/repos/tdwg/bdq/issues/50
AMENDMENT_COORDINATES_FROM_VERBATIM	Propose amendment to the values of dwc:decimalLatitude and dwc:decimalLongitude from information in the verbatim coordinates terms.	https://api.github.com/repos/tdwg/bdq/issues/32
VALIDATION_COORDINATES_NOTZERO	Are the values of either dwc:decimalLatitude or dwc:decimalLongitude numbers that are not equal to 0?	https://api.github.com/repos/tdwg/bdq/issues/87
VALIDATION_COORDINATES_STATEPROVINCE_CONSISTENT	Do the geographic coordinates fall on or within the boundary from the bdq:sourceAuthority for the given dwc:stateProvince or within the distance given by bdq:spatialBufferInMeters outside that boundary?	https://api.github.com/repos/tdwg/bdq/issues/56
VALIDATION_COORDINATES_TERRESTRIALMARINE	Does the marine/non-marine biome of a taxon from the bdq:sourceAuthority match the biome at the location given by the coordinates?	https://api.github.com/repos/tdwg/bdq/issues/51
AMENDMENT_COORDINATES_TRANSPOSED	Propose amendment of the signs of dwc:decimalLatitude and/or dwc:decimalLongitude to align the location with the dwc:countryCode.	https://api.github.com/repos/tdwg/bdq/issues/54
VALIDATION_COORDINATEUNCERTAINTY_INRANGE	Is the value of dwc:coordinateUncertaintyInMeters a number between 1 and 20,037,509?	https://api.github.com/repos/tdwg/bdq/issues/109
VALIDATION_COUNTRY_COUNTRYCODE_CONSISTENT	Does the ISO country code determined from the value of dwc:country equal the value of dwc:countryCode?	https://api.github.com/repos/tdwg/bdq/issues/62
VALIDATION_COUNTRY_FOUND	Does the value of dwc:country occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/21
VALIDATION_COUNTRY_NOTEMPTY	Is there a value in dwc:country?	https://api.github.com/repos/tdwg/bdq/issues/42
AMENDMENT_COUNTRYCODE_FROM_COORDINATES	Propose amendment to the value of dwc:countryCode if dwc:decimalLatitude and dwc:decimalLongitude fall within a boundary from the bdq:sourceAuthority that is attributable to a single valid country code.	https://api.github.com/repos/tdwg/bdq/issues/73
VALIDATION_COUNTRYCODE_NOTEMPTY	Is there a value in dwc:countryCode?	https://api.github.com/repos/tdwg/bdq/issues/98
VALIDATION_COUNTRYCODE_STANDARD	Is the value of dwc:countryCode a valid ISO 3166-1-alpha-2 country code?	https://api.github.com/repos/tdwg/bdq/issues/20
AMENDMENT_COUNTRYCODE_STANDARDIZED	Propose amendment to the value of dwc:countryCode if it can be interpreted as an ISO country code.	https://api.github.com/repos/tdwg/bdq/issues/48
VALIDATION_COUNTRYSTATEPROVINCE_CONSISTENT	Are the combination of the values of dwc:country, dwc:stateProvince consistent with the values in the bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/200
VALIDATION_COUNTRYSTATEPROVINCE_UNAMBIGUOUS	Is the combination of the values of the terms dwc:country, dwc:stateProvince unique in the bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/201
ISSUE_DATAGENERALIZATIONS_NOTEMPTY	Is there a value in dwc:dataGeneralizations?	https://api.github.com/repos/tdwg/bdq/issues/72
VALIDATION_DATEIDENTIFIED_INRANGE	Is the value of dwc:dateIdentified within Parameter ranges and either overlap or is later than dwc:eventDate?	https://api.github.com/repos/tdwg/bdq/issues/76
VALIDATION_DATEIDENTIFIED_STANDARD	Is the value of dwc:dateIdentified a valid ISO date?	https://api.github.com/repos/tdwg/bdq/issues/69
AMENDMENT_DATEIDENTIFIED_STANDARDIZED	Propose amendment to the value of dwc:dateIdentified to a valid ISO date.	https://api.github.com/repos/tdwg/bdq/issues/26
VALIDATION_DAY_INRANGE	Is the value of dwc:day interpretable as a valid integer between 1 and 28 inclusive or 29, 30 or 31 given the relative month and year?	https://api.github.com/repos/tdwg/bdq/issues/125
VALIDATION_DAY_STANDARD	Is the value of dwc:day an integer between 1 and 31 inclusive?	https://api.github.com/repos/tdwg/bdq/issues/147
AMENDMENT_DAY_STANDARDIZED	Propose amendment to the value of dwc:day as an integer between 1 and 31 inclusive.	https://api.github.com/repos/tdwg/bdq/issues/127
VALIDATION_DCTYPE_NOTEMPTY	Is there a value in dc:type?	https://api.github.com/repos/tdwg/bdq/issues/103
VALIDATION_DCTYPE_STANDARD	Does the value in dc:type occur as a value in the DCMI type vocabulary?	https://api.github.com/repos/tdwg/bdq/issues/91
AMENDMENT_DCTYPE_STANDARDIZED	Propose amendment to the value of dc:type using the DCMI type vocabulary.	https://api.github.com/repos/tdwg/bdq/issues/41
VALIDATION_DECIMALLATITUDE_INRANGE	Is the value of dwc:decimalLatitude a number between -90 and 90 inclusive?	https://api.github.com/repos/tdwg/bdq/issues/79
VALIDATION_DECIMALLATITUDE_NOTEMPTY	Is there a value in dwc:decimalLatitude?	https://api.github.com/repos/tdwg/bdq/issues/119
VALIDATION_DECIMALLONGITUDE_INRANGE	Is the value of dwc:decimalLongitude a number between -180 and 180 inclusive?	https://api.github.com/repos/tdwg/bdq/issues/30
VALIDATION_DECIMALLONGITUDE_NOTEMPTY	Is there a value in dwc:decimalLongitude?	https://api.github.com/repos/tdwg/bdq/issues/96
VALIDATION_ENDDAYOFYEAR_INRANGE	Is the value of dwc:endDayOfYear an integer between 1 and 365 inclusive, or 366 if a leap year?	https://api.github.com/repos/tdwg/bdq/issues/131
ISSUE_ESTABLISHMENTMEANS_NOTEMPTY	Is there a value in dwc:establishmentMeans?	https://api.github.com/repos/tdwg/bdq/issues/94
AMENDMENT_EVENT_FROM_EVENTDATE	Propose amendment to values in any of dwc:year, dwc:month, dwc:day, dwc:startDayOfYear or dwc:endDayOfYear from the content of dwc:eventDate.	https://api.github.com/repos/tdwg/bdq/issues/52
VALIDATION_EVENT_TEMPORAL_NOTEMPTY	Is there a value in any of the terms dwc:eventDate, dwc:year, dwc:month, dwc:day, dwc:startDayOfYear, dwc:endDayOfYear, dwc:verbatimEventDate?	https://api.github.com/repos/tdwg/bdq/issues/88
VALIDATION_EVENT_CONSISTENT	Are the values in dwc:eventDate consistent with the values in dwc:year, dwc:month, dwc:day, dwc:startDayOfYear and dwc:endDayOfYear?	https://api.github.com/repos/tdwg/bdq/issues/67
MEASURE_EVENTDATE_DURATIONINSECONDS	What is the duration of dwc:eventDate in seconds?	https://api.github.com/repos/tdwg/bdq/issues/140
AMENDMENT_EVENTDATE_FROM_VERBATIM	Propose amendment to the value of dwc:eventDate from the content of dwc:verbatimEventDate.	https://api.github.com/repos/tdwg/bdq/issues/86
AMENDMENT_EVENTDATE_FROM_YEARMONTHDAY	Propose amendment to the value of dwc:eventDate from values in dwc:year, dwc:month and dwc:day.	https://api.github.com/repos/tdwg/bdq/issues/93
AMENDMENT_EVENTDATE_FROM_YEARSTARTDAYOFYEARENDDAYOFYEAR	Propose amendment to the value of dwc:eventDate from values in dwc:year, dwc:startDayOfYear and dwc:endDayOfYear.	https://api.github.com/repos/tdwg/bdq/issues/132
VALIDATION_EVENTDATE_INRANGE	Is the value of dwc:eventDate entirely with the Parameter Range?	https://api.github.com/repos/tdwg/bdq/issues/36
VALIDATION_EVENTDATE_NOTEMPTY	Is there a value in dwc:eventDate?	https://api.github.com/repos/tdwg/bdq/issues/33
VALIDATION_EVENTDATE_STANDARD	Is the value of dwc:eventDate a valid ISO date?	https://api.github.com/repos/tdwg/bdq/issues/66
AMENDMENT_EVENTDATE_STANDARDIZED	Propose amendment of the value of dwc:eventDate to a valid ISO date.	https://api.github.com/repos/tdwg/bdq/issues/61
VALIDATION_FAMILY_FOUND	Does the value of dwc:family occur at rank of Family in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/28
VALIDATION_GENUS_FOUND	Does the value of dwc:genus occur at the rank of Genus in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/122
AMENDMENT_GEODETICDATUM_ASSUMEDDEFAULT	Propose amendment to dwc:geodeticDatum using the value of bdq:defaultGeodeticDatum if dwc:geodeticDatum is empty. If dwc:coordinateUncertaintyInMeters is not empty and there are not empty values for dwc:latitude and dwc:longitude, amend dwc:coordinateUncertaintyInMeters by adding a maximum datum shift.	https://api.github.com/repos/tdwg/bdq/issues/102
VALIDATION_GEODETICDATUM_NOTEMPTY	Is there a value in dwc:geodeticDatum?	https://api.github.com/repos/tdwg/bdq/issues/78
VALIDATION_GEODETICDATUM_STANDARD	Does the value of dwc:geodeticDatum occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/59
AMENDMENT_GEODETICDATUM_STANDARDIZED	Propose amendment to the value of dwc:geodeticDatum using bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/60
VALIDATION_KINGDOM_FOUND	Does the value of dwc:kingdom occur at rank of Kingdom in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/81
VALIDATION_LICENSE_NOTEMPTY	Is there a value in dcterms:license?	https://api.github.com/repos/tdwg/bdq/issues/99
VALIDATION_LICENSE_STANDARD	Does the value of dcterms:license occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/38
AMENDMENT_LICENSE_STANDARDIZED	Propose amendment to the value of dwc:license using bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/133
VALIDATION_LOCATION_NOTEMPTY	Is there a value in any of the Darwin Core spatial terms that could specify a location?	https://api.github.com/repos/tdwg/bdq/issues/40
VALIDATION_MAXDEPTH_INRANGE	Is the value of dwc:maximumDepthInMeters within the Parameter range?	https://api.github.com/repos/tdwg/bdq/issues/187
VALIDATION_MAXELEVATION_INRANGE	Is the value of dwc:maximumElevationInMeters of a single record within a valid range	https://api.github.com/repos/tdwg/bdq/issues/112
VALIDATION_MINDEPTH_INRANGE	Is the value of dwc:minimumDepthInMeters within the Parameter range?	https://api.github.com/repos/tdwg/bdq/issues/107
VALIDATION_MINDEPTH_LESSTHAN_MAXDEPTH	Is the value of dwc:minimumDepthInMeters a number that is less than or equal to the value of dwc:maximumDepthInMeters?	https://api.github.com/repos/tdwg/bdq/issues/24
AMENDMENT_MINDEPTH-MAXDEPTH_FROM_VERBATIM	Propose amendments of the values of dwc:minimumDepthInMeters and/or dwc:maximumDepthInMeters if they can be interpreted from dwc:verbatimDepth.	https://api.github.com/repos/tdwg/bdq/issues/55
VALIDATION_MINELEVATION_INRANGE	Is the value of dwc:minimumElevationInMeters within the Parameter range?	https://api.github.com/repos/tdwg/bdq/issues/39
VALIDATION_MINELEVATION_LESSTHAN_MAXELEVATION	Is the value of dwc:minimumElevationInMeters a number less than or equal to the value of dwc:maximumElevationInMeters?	https://api.github.com/repos/tdwg/bdq/issues/108
AMENDMENT_MINELEVATION-MAXELEVATION_FROM_VERBATIM	Propose amendment(s) to the values of dwc:minimumElevationInMeters and/or dwc:maximumElevationInMeters if they can be interpreted from dwc:verbatimElevation.	https://api.github.com/repos/tdwg/bdq/issues/68
VALIDATION_MONTH_STANDARD	Is the value of dwc:month interpretable as an integer between 1 and 12 inclusive?	https://api.github.com/repos/tdwg/bdq/issues/126
AMENDMENT_MONTH_STANDARDIZED	Propose an amendment to the value of dwc:month as an integer between 1 and 12 inclusive.	https://api.github.com/repos/tdwg/bdq/issues/128
VALIDATION_OCCURRENCEID_NOTEMPTY	Is there a value in dwc:occurrenceID?	https://api.github.com/repos/tdwg/bdq/issues/47
VALIDATION_OCCURRENCEID_STANDARD	Does the value of dwc:occurrenceID occur in bdqSourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/23
AMENDMENT_OCCURRENCESTATUS_ASSUMEDDEFAULT	Propose an amendment of the value of dwc:occurrenceStatus to the default parameter value if dwc:occurrenceStatus, dwc:individualCount and dwc:organismQuantity are empty.	https://api.github.com/repos/tdwg/bdq/issues/75
VALIDATION_OCCURRENCESTATUS_NOTEMPTY	Is there a value in dwc:occurrenceStatus?	https://api.github.com/repos/tdwg/bdq/issues/117
VALIDATION_OCCURRENCESTATUS_STANDARD	Does the value of dwc:occurrenceStatus occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/116
AMENDMENT_OCCURRENCESTATUS_STANDARDIZED	Propose amendment to the value of dwc:occurrenceStatus using bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/115
VALIDATION_ORDER_FOUND	Does the value of dwc:order occur at rank of Order in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/83
VALIDATION_PHYLUM_FOUND	Does the value of dwc:phylum occur at rank of Phylum in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/22
VALIDATION_POLYNOMIAL_CONSISTENT	Is the polynomial represented in dwc:scientificName consistent with the equivalent values in dwc:genericName, dwc:specificEpithet, dwc:infraspecificEpithet?	https://api.github.com/repos/tdwg/bdq/issues/101
VALIDATION_SCIENTIFICNAME_FOUND	Is there a match of the contents of dwc:scientificName with bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/46
AMENDMENT_SCIENTIFICNAME_FROM_TAXONID	Propose an amendment to the value of dwc:scientificName using the taxonID value from bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/71
VALIDATION_SCIENTIFICNAME_NOTEMPTY	Is there a value in dwc:scientificName?	https://api.github.com/repos/tdwg/bdq/issues/82
VALIDATION_STARTDAYOFYEAR_INRANGE	Is the value of dwc:startDayOfYear an integer between 1 and 365 inclusive, or 366 if a leap year?	https://api.github.com/repos/tdwg/bdq/issues/130
VALIDATION_STATEPROVINCE_FOUND	Does the value of dwc:stateProvince occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/199
VALIDATION_TAXON_NOTEMPTY	Is there a value in any of the terms needed to determine that the taxon exists?	https://api.github.com/repos/tdwg/bdq/issues/105
VALIDATION_TAXON_UNAMBIGUOUS	Can the taxon be unambiguously resolved from bdq:sourceAuthority using the available taxon terms?	https://api.github.com/repos/tdwg/bdq/issues/70
VALIDATION_TAXONID_COMPLETE	Does the value of dwc:taxonID contain a complete identifier?	https://api.github.com/repos/tdwg/bdq/issues/121
AMENDMENT_TAXONID_FROM_TAXON	Propose amendment to the value of dwc:taxonID if it can be unambiguously resolved from bdq:sourceAuthority using the available taxon terms.	https://api.github.com/repos/tdwg/bdq/issues/57
VALIDATION_TAXONID_NOTEMPTY	Is there a value in dwc:taxonID?	https://api.github.com/repos/tdwg/bdq/issues/120
VALIDATION_TAXONRANK_NOTEMPTY	Is there a value in dwc:taxonRank?	https://api.github.com/repos/tdwg/bdq/issues/161
VALIDATION_TAXONRANK_STANDARD	Does the value of dwc:taxonRank occur in bdq:sourceAuthority?	https://api.github.com/repos/tdwg/bdq/issues/162
AMENDMENT_TAXONRANK_STANDARDIZED	Propose amendment to the value of dwc:taxonRank using bdq:sourceAuthority.	https://api.github.com/repos/tdwg/bdq/issues/163
MEASURE_VALIDATIONTESTS_COMPLIANT	The number of distinct VALIDATION tests that have a Response.status="COMPLIANT" for a given record.	https://api.github.com/repos/tdwg/bdq/issues/135
MEASURE_VALIDATIONTESTS_NOTCOMPLIANT	The number of distinct VALIDATION tests that have a Response.status="NOT_COMPLIANT" for a given record.	https://api.github.com/repos/tdwg/bdq/issues/31
MEASURE_VALIDATIONTESTS_PREREQUISITESNOTMET	The number of distinct VALIDATION tests that have a Response.status="EXTERNAL_PREREQUISITES_NOT_MET" or "INTERNAL_PREREQUISITES_NOT_MET" for a given record.	https://api.github.com/repos/tdwg/bdq/issues/134
VALIDATION_YEAR_INRANGE	Is the value of dwc:year within the Parameter range?	https://api.github.com/repos/tdwg/bdq/issues/84
VALIDATION_YEAR_NOTEMPTY	Is there a value in dwc:year?	https://api.github.com/repos/tdwg/bdq/issues/49

6.5 MultiRecord Measure Specifications (Normative)

[Include generated text]

7 Validating test implementations (Lee)
Test data…and something needed about PROCEDURE for updating test data if relevant NORMATIVE terms change! See email 11/6/2023 from Lee back to all on Pau’s issue with DataID 320.

7.1 Introduction

7.2 Example data for validating tests (Informative)
SUPP 3

7.3 Identifying example data (Normative) (Paul)

8.0 Acknowledgements
Alex Thompson, Paulo Zermoglio, Antonio Mauro Saraiva, Allan Koch Veiga, Tim Roberston, Yi-Min Gan 
9.0 References

Chapman AD, Belbin L, Zermoglio PF, Wieczorek J, Morris PJ, Nicholls M, Rees ER, Veiga AK, Thompson A, Saraiva AM, James SA, Gendreau C, Benson A, Schigel D (2020). Developing Standards for Improved Data Quality and for Selecting Fit for Use Biodiversity Data. Biodiversity Information Science and Standards 4: e50889. https://doi.org/10.3897/biss.4.50889
Veiga AK, Saraiva AM, Chapman AD, Morris PJ, Gendreau C, Schigel D, and Robertson TJ (2017). A conceptual framework for quality assessment and management of biodiversity data. PLOS ONE 12(6): e0178731. https://doi.org/10.1371/journal.pone.0178731

Darwin Core

**Comment by Tasilee** [comment]
![Capture](https://github.com/tdwg/bdq/assets/6656350/cafd85f3-df32-4511-a06d-904609f27dba)

**Comment by chicoreus** [comment]
@Tasilee order does matter for a few Amendments.

Draft section 5.1.3:  

5.1.3. Amendments where order is important (Normative)

When Amendments are executed in a workflow where downstream Amendments operate on data with the changes proposed by upstream Amendments applied, the following sequences SHOULD be followed.   Similarly when Amendments are executed in parallel these sequences SHOULD be applied.

Give amendments propose a value to a primary term from secondary terms priority over those which back fill secondary terms from a primary term, 
AMENDMENT_EVENT_FROM_EVENTDATE SHOULD be run after the following Ammendments that propose changes to dwc:eventDate:  AMENDMENT_EVENTDATE_FROM_VERBATIM, 
AMENDMENT_EVENTDATE_FROM_YEARMONTHDAY, 
AMENDMENT_EVENTDATE_FROM_YEARSTARTDAYOFYEARENDDAYOFYEAR, 
AMENDMENT_EVENTDATE_STANDARDIZED.

AMENDMENT_SCIENTIFICNAME_FROM_TAXONID SHOULD be run after this Amendment which proposes changes to dwc:TaxonID: AMENDMENT_TAXONID_FROM_TAXON

Where multiple Amendments on secondary terms could propose conflicting changes to a primary term, the sequence of Amendments SHOULD be ordered.

The following Amendments SHOULD be composed to run in an ordered sequence: 
first, AMENDMENT_EVENTDATE_FROM_VERBATIM, second, AMENDMENT_EVENTDATE_FROM_YEARSTARTDAYOFYEARENDDAYOFYEAR and finally AMENDMENT_EVENTDATE_FROM_YEARMONTHDAY,

**Comment by Tasilee** [comment]
Thanks @chicoreus - that makes sense. 

Regarding a shared area where we can all see the standard doc, I haven't a clue on what could substitute for Google Docs. Any ideas?

**Comment by chicoreus** [comment]
On Fri, 23 Jun 2023 19:03:33 -0700
Lee Belbin ***@***.***> wrote:

> Thanks @chicoreus - that makes sense. 
> 
> Regarding a shared area where we can all see the standard doc, I
> haven't a clue on what could substitute for Google Docs. Any ideas?

The logical place for long term maintinance would be a markdown document in the tdwg/bdq github repository.  Similtaneous colaborative editing wouldn't be possible, but edits in forks with pull requests and merges of those serving as an editorial process would be supported.

**Comment by Tasilee** [comment]
Using github for the document would be good, but I'd suggest that option when we have a final draft. There will be a lot of changes over the next month or so.

**Comment by ArthurChapman** [comment]
I agree - GitHub is not an ideal platform to write a document.  What is wrong with Google Docs?  We all use it and it works well.

**Comment by Tasilee** [comment]
The standards document is now in github

**Comment by chicoreus** [comment]
Some notes from TG2 call: 

Validations are independent, internally order does not matter.

Some amendments are interdependent and order in which they are run (or the ordering for resolving conflicts) matters.

Most interdependencies are resolved by executing all validations (in parallel or in sequence), then when these are completed all amendments run (in parallel or in sequence, but with a few key interdependencies), then when these are completed run all validations (in parallel or in sequence) (or validations run where amendments proposed changes), with the second round of validations run on the data with proposed amendments applied.   Comparing the validations pre-amendment and post-amendment gives an assessment of the data quality (for core purposes), and how much that quality would be improved by accepting all of the changes proposed by the amendments.

**Comment by chicoreus** [comment]
Minimal safe workflows, when passing SingleRecords through a data processing pipeline: 

(1) Run all validations, in sequence, or in parallel.

(2) Run all amendments, with a few having a required sequence (otherwise in sequence or in parallel), then when complete, run all validations (in sequence or in parallel), with proposed changes from amendments applied to the data.

(3)  Run all validations, in sequence, or in parallel, then when these are completed, Run all amendments, with a few having a required sequence (otherwise in sequence or in parallel), then when complete, run all validations (in sequence or in parallel), with proposed changes from amendments applied to the data.

Workflows that combine multiple pipelines of amendments and validations, need to be cognizant of information elements shared among tests and of the order dependence of a small set of annotations.

When operating on distinct values in a MultiRecord, attention needs to be paid to test interdependencies.
